README.md

# *VUE 3 JOURNEY JUST STARTED*

<div align="center">
  <img width="318" heigth="256" src="https://raw.githubusercontent.com/tvjsx/trading-vue-js/master/assets/README-c8a97eb7.png?raw=true" alt="trading-vue logo">
</div>

# TradingVue.js ![npm](https://img.shields.io/npm/v/trading-vue-js.svg?color=brightgreen&label=version) ![license](https://img.shields.io/badge/license-MIT-blue.svg) ![build](https://img.shields.io/badge/build-passing-brightgreen.svg) ![size](https://img.shields.io/github/size/tvjsx/trading-vue-js/dist/trading-vue.min.js.svg) [![Gitter](https://badges.gitter.im/Trading-Vue-js/community.svg)](https://gitter.im/Trading-Vue-js/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge) [![DONATE](https://img.shields.io/badge/฿-Donate-f7921b.svg)](https://github.com/tvjsx/trading-vue-js/blob/master/DONATION.md)

**TradingVue.js** is a hackable charting lib for traders. You can draw literally ANYTHING on top of candlestick charts.

<div align="center">
<br>
  <a href="https://discord.gg/PKD4PUy"> <img src="https://raw.githubusercontent.com/tvjsx/trading-vue-js/master/assets/README-06bc5a15.png" alt="trading-vue logo"></a>
</div>


## Why

If you create trading software - this lib is probably for you. If you like to make custom indicators and think out of the box - this lib is most likely for you. And if you miss usability of [TradingView.com](https://www.tradingview.com) in other open-source libraries and can't stand it - you are definetly in the right place!

<br>

<div align="center">
    <img width="780" heigth="446" src="https://github.com/tvjsx/trading-vue-js/blob/master/assets/README-new-preview.gif?raw=true" alt="White preview">
</div>

<br>

## Features

* Scrolling & zooming as we all like
* Simple API for making new overlays
* Custom drawing tools
* Non-time based charts (e.g. Renko)
* One overlay === one .vue component
* Fully reactive
* Fully responsive
* Customizable colors and fonts
* Quite fast (works even with 3 mil candles)
* Scripts (make your own indicators)

## Demo & Docs

### [Demo](https://tvjsx.github.io/trading-vue-demo/) | [Getting Started](https://github.com/tvjsx/trading-vue-js/tree/master/docs/guide) | [API Book](https://github.com/tvjsx/trading-vue-js/tree/master/docs/api#api-book) | [Built-in Overlays](https://github.com/tvjsx/trading-vue-js/tree/master/docs/overlays#built-in-overlays) | [Examples](https://github.com/tvjsx/trading-vue-js/tree/master/test/tests) | [101 Project](https://github.com/tvjsx/trading-vue-101) | [llll Gitter](https://gitter.im/Trading-Vue-js) | [FAQ](https://github.com/tvjsx/trading-vue-js/tree/master/docs/faq) | [FREE Overlay Pack](https://github.com/tvjsx/tvjs-overlays) | [Free XP Pack](https://github.com/tvjsx/tvjs-xp) | [Playground](https://tvjs.io/play)

*To run the examples, download the repo & `npm run test`*

## Install

**NPM**
```
npm i trading-vue-js
```
**In browser**

```
<script src="trading-vue.min.js"></script>
```

## How to use

Minimal working example:

```html
<template>
<trading-vue :data="this.$data"></trading-vue>
</template>
<script>

import TradingVue from 'trading-vue-js'

export default {
    name: 'app',
    components: { TradingVue },
    data() {
        return {
            ohlcv: [
                [ 1551128400000, 33,  37.1, 14,  14,  196 ],
                [ 1551132000000, 13.7, 30, 6.6,  30,  206 ],
                [ 1551135600000, 29.9, 33, 21.3, 21.8, 74 ],
                [ 1551139200000, 21.7, 25.9, 18, 24,  140 ],
                [ 1551142800000, 24.1, 24.1, 24, 24.1, 29 ],
            ]
        }
    }
}

</script>
```

## Core philosophy

The core philosophy is **Data -> Screen Mapping (DSM)**. The library provides you with functions that map your data (it could be anything) to screen coordinates. The lib does all the dirty work behind the scenes: scrolling, scaling, reactivity, etc.

 ```
 layout.t2screen(t) // time -> x
 layout.$2screen($) // price -> y
 layout.t_magnet(t) // time -> nearest candle x
 layout.screen2$(y) // y -> price
 layout.screen2t(x) // x -> time
 ```
 Using these functions and the standard js canvas API, you can do magic.


## Data structure

PRO TIP: **chart** is mandatory if you want to see something other than a white screen

IMPORTANT: All data must be sorted by time (in ascending order). The main OHLCV must not contain duplicate timestamps.

[Full version of DataStructure](https://github.com/tvjsx/trading-vue-js/tree/master/docs/api#data-structure-new)

```js
{
    "chart": {   // Mandatory
        "type": "<Chart Type, e.g. Candles>",
        "data": [
            [timestamp, open, high, low, close, volume],
            ...
        ],
        "settings": { } // Settings (depending on Chart Type)
    },
    "onchart": [ // Displayed ON the chart
        {
            "name": "<Indicator name>",
            "type": "<e.g. EMA, SMA>",
            "data": [
                [timestamp, ... ], // Arbitrary length
                ...
            ],
            "settings": { } // Arbitrary settings format
        },
        ...
    ],
    "offchart": [ // Displayed BELOW the chart
        {
            "name": "<Indicator name>",
            "type": "<e.g. RSI, Stoch>",
            "data": [
                [timestamp, ... ], // Arbitrary length
                ...
            ],
            "settings": { } // Arbitrary settings format
        },
        ...
    ]
}
```
The process of adding a new indicator is simple: first you define your own data format (should be timestamped though) and display settings. For example, EMA data might look like this:

 ```json
 {
     "name": "EMA, 25",
     "type": "EMA",
     "data": [
         [ 1551128400000, 3091 ],
         [ 1551132000000, 3112 ],
         [ 1551135600000, 3105 ]
     ],
     "settings": {
         "color": "#42b28a"
     }
 },

 ```

## Example of a simple overlay class

And then you make a new overlay class to display that data on the grid:

```js
import { Overlay } from 'trading-vue-js'

export default {
    name: 'EMA',
    mixins: [Overlay],
    methods: {
        draw(ctx) {
            const layout = this.$props.layout
            ctx.strokeStyle = this.color
            ctx.beginPath()

            for (var p of this.$props.data) {

                // t2screen & $2screen - special functions that
                // map your data coordinates to grid coordinates
                let x = layout.t2screen(p[0])
                let y = layout.$2screen(p[1])

                ctx.lineTo(x, y)
            }

            ctx.stroke()
        },
        use_for() { return ['EMA'] },
        data_colors() { return [this.color] }
    },
    computed: {
        color() {
            return this.$props.settings.color
        }
    }
}
```

That's why the title doesn't lie: you can draw ANYTHING.

## Grin

<div align="center">
  <img src="https://github.com/tvjsx/trading-vue-js/blob/master/assets/GRIN.gif?raw=true" alt="trading-vue logo">
</div>

#### [ Code | click your 🐁 ](https://gist.github.com/C451/6d8ea8d603e3faadd801b4b3f2e7a57d)


## Roadmap

* ~~Docs~~
* ~~Tests~~
* Solve known issues (marked as 'TODO: IMPORTANT')
* Performance improvements
* ~~Data-manipulation helpers~~
* ~~Add more built-in overlays~~
* ~~Add toolbar (drawing tools)~~
* Custom loayout / layout persistence
* Fix and improve mobile version
* **Version 1.0.0 here**

Progress in details: https://github.com/tvjsx/trading-vue-js/projects/1

## Changelog

See CHANGELOG.md

## Development & Building

**Install devDependencies**

```
npm install
```

**Run development enviroment (hot)**

```
npm run dev
```
*Server is running on http://localhost:8080*

**Run in a CDN-like mode**

```
npm run cdn
```
*Server is running on http://localhost:8080*

**Build the bundle**

```
npm run build
```

**Visual testing**

```
npm run test
```

*Server is running on http://localhost:8080*


**Automatic testing**

```
npm run auto-test
```


## Contributing

1. Fork ( https://github.com/tvjsx/trading-vue-js/fork )
2. Create your feature branch (`git checkout -b cool-new-feature`)
3. Commit your changes (`git commit -am 'Let's rock smth'`)
4. Push to the branch (`git push origin cool-new-feature`)
5. Create a new Pull Request

Please read the guidelines in CONTRIBUTING.md


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.10.0] - 2020-12-07

- The first package of mobile improvements
- Enhances skins
- Small modifications of scripts execution
- Improvements of Interfaces (aka inline overlay UI)
- Script fixes (update onchart/offchart OVs)
- New aggtype: copy (fast aggregation through copying)
- CSS tranforms support (transition)
- Overlay name templates, e.g. "RSI, $length"
- Fixed loading animation for some cases

## [0.9.0] - 2020-11-17

- Scripts: advanced samplers (sym() function)
- Implemented Dataset ops (update, merge, remove)
- Countless small bug fixes as always
- Scripts: add onclose()
- Fixed auto-scroll (when new candle is added)
- Scripts: offchart/onchart updates
- DC: aggtool v2 (completely rewritten)
- Custom legend buttons (your custom icon)
- Tool groups
- Scroll wheel modes: prevent, pass, click
- WebWorker RAM limit
- LineTool: add Ray mode
- Legend buttons handlers [EXT]

## [0.8.0] - 2020-10-15

- Timezones (initial support)
- Extended shader props
- Show/hide scalebars (via chart-config sizes)
- Chart section resizing [Ext]
- Linked charts [Ext]
- Shift + drag => Measure
- Various bug kills and memory leak demolitions
- Scripts: onrange exec, node.js SE exec (opt)
- Scripts: added support of 'tf', 'range'
- Updated RangeTool
- Datasets, very first try
- 30% build size decrease

## [0.7.0] - 2020-9-15
- Extensions (widgets, colorpacks & skins v1)
- Pins & toolbar bug fixes
- WebWorker modules (emitted from extension)
- Std lib injections (though WW modules)
- Improved fn argument parsing for scripts
- Splines overlay: skipNaN mode
- Scripts: multi-tf TA, how cool is that?
- Scripts: hl2, hlc3, ohlc4 and potentially more in the futr
- Script: SE hooks for backtester mod (if you like)
- Hooks on the internal events (now only for Chart.vue)
- Fixed legend values (when grid.id is custom)
- DC: some improvements & new settings
- lz-string compression for the WW code. Supr fast

## [0.6.0] - 2020-8-15
- Scripts (early state)
- DC faster update()
- DC tick aggregation
- Improved offchart overlayz performance
- Multiple bug fixes
- +30 Std overlayz (MACD, DMI, Stoch ...)
- Overlay loading state (animation)
- TradingLite-like zoom mode
- Better IB mode + Illuminati test

## [0.5.0] - 2020-5-5
### Changes
- Index-based rendering mode (hides weekend gaps + sup Renko)
- Log-scale mode
- Interfaces (preview)
- Performance fixes
- Zoom to the current candle mode
- Pixel-perfect candles, candle wicks
- Offchart drawing tools
- Price level
- Custom grid heights
- Sidear & Botbar inline shaders
- Added RangeTool
- Forced chart timeframe (overwrites the auto-detected one)
- Updated DataCube

## [0.4.0] - 2019-11-2
### Changes
- Multiple bug fixes and improvements
- Legend buttons
- New/updated overlays (Segment, Splitters, Candles)
- Data Structure v1.1
- Mac support (Trackpad + Retina)
- New API functions: setRange(), goto(), getCursor() etc.
- Better timeline
- New interval detection algo
- DataCube
- Tool overlay extension
- Drawing tools (Line, Extended Line)

## [0.3.0] - 2019-5-17
### Changes
- Multiple bug fixes (legend, slicelib, cursor)
- New overlays: Trades, Channel, Splines
- Added visual testing suite
- Mouse events for overlays
- Several fixes that made charts sharper and slicker
- Hackable chart config. Heck, how can the most hackable lib be without it?

## [0.2.0] - 2019-3-27
### Changes
- Added y-range transform for overlays
- Performance boost with fancy slice lib
- Implemented z-index ordering
- Vertical scaling implementation v1
- Added data masking feature, now you can include strings
- Several smaller fixes


tools/og_data.js
// OG data secured

const request = require('request')
const tulind = require('tulind')
const I = tulind.indicators
const fs = require('fs')

const write_to = './data/data.json'

const t1 = Math.floor(new Date().getTime() / 1000)
const t0 = t1 - 3600 * 4200

// Hey Coingy, we will borrow some data, ok?
const mrk = 'BITF/BTC/USD'
const url = `https://www.coinigy.com/getjson/chart_feed/${mrk}/60/${t0}/${t1}`

request
    .get(url, {}, function(err,res,body) {
        if (err) {
            console.log('Not this time, bro')
            process.exit()
        } else {
            try {
                deal_with_it(JSON.parse(body))
            } catch(e) {
                console.log('Data corrupted:', e)
            }
        }
    })

async function deal_with_it(data) {

    data = data.map(x => {
        return [
            x[0] * 1000,
            parseFloat(x[1]),
            parseFloat(x[2]),
            parseFloat(x[3]),
            parseFloat(x[4]),
            parseFloat(x[5]),
        ]
    })

    data.reverse()

    data = {
        ohlcv: data,
        onchart: [
            {
                name: "EMA, 25",
                type: "EMA",
                data: await make_ema(data, 25),
                settings: {}
            },
            {
                name: "EMA, 43",
                type: "EMA",
                data: await make_ema(data, 43),
                settings: {}
            },
            {
                name: "Trades",
                type: "Trades",
                data: await trade_like_god(data),
                settings: {
                    "z-index": 5,
                    "legend": false
                }
            }

        ],
        offchart: [
            {
                name: "RSI, 20",
                type: "RSI",
                data: await make_rsi(data, 20),
                settings: {
                    upper: 70,
                    lower: 30,
                    backColor: "#9b9ba316",
                    bandColor: "#666"
                }
            }
        ]
    }

    save(data)
}

async function make_ema(data, len) {
    return new Promise(rs => {
        const close = data.map(x => x[4])
        const time = data.map(x => x[0])
        I.ema.indicator([close], [len], (err, results) => {
            rs(merge(time, results[0]))
        })
    })
}

async function make_rsi(data, len) {
    return new Promise(rs => {
        const close = data.map(x => x[4])
        const time = data.map(x => x[0])
        I.rsi.indicator([close], [len], (err, results) => {
            rs(merge(time, results[0]))
        })
    })
}

async function trade_like_god(data, len) {
    let trades = []
    let entry = 0
    let last = [0, 0]
    return new Promise(rs => {
        for (var p of data) {
            if (Math.random() < 0.1 && last[1] === 0) {
                entry = (p[2] -  p[3]) * Math.random() + p[3]
                last = [p[0], 1, entry]
                trades.push(last)
                continue
            }
            if (Math.random() < 0.1 && last[1] === 1) {
                let sell = (p[2] -  p[3]) * Math.random() + p[3]
                if (sell > entry) {
                    let pr = '+' + (100 * (sell / entry - 1)).toFixed(2) + '%'
                    last = [p[0], 0, sell, pr]
                    trades.push(last)
                }
                if (p[3] / entry < 0.999) {
                    trades.pop()
                    last = [0,0]
                }
            }
        }
        rs(trades)
    })
}


// Merge backwards
function merge(vec1, vec2) {
    const len = Math.min(vec1.length, vec2.length)
    let vec = []
    for (var i = 0; i < len; i++) {
        let i1 = vec1.length - i - 1
        let i2 = vec2.length - i - 1
        vec.unshift([vec1[i1], vec2[i2]])
    }
    return vec
}

function save(data) {

    fs.writeFileSync(write_to, JSON.stringify(data,null,4))

    console.log('Data secured (4200 points)')
}
